lib/Sema/TypeCheckConstraints.cpp

```diff
@@ -2175,7 +2176,8 @@ Type TypeChecker::typeCheckExpressionImpl(Expr *&expr, DeclContext *dc,
   // Construct a constraint system from this expression.
   ConstraintSystemOptions csOptions = ConstraintSystemFlags::AllowFixes;
 
-  if (options.contains(TypeCheckExprFlags::SuppressDiagnostics))
+  if (options.contains(TypeCheckExprFlags::SuppressDiagnostics) ||
+      DiagnosticSuppression::isEnabled(Diags))
```

Comment:
@xedin Do we really need both mechanisms?

Reply:
@rintaro Shouldn't enabled diagnostic suppression mean SuppressDiagnostics flag while forming constraint system?

Reply:
@xedin 
Currently, `DiagnosticSuppression` does nothing related to `TypeCheckExprFlags::SuppressDiagnostics`.

@slavapestov 
We can drop `TypeCheckExprFlags::SuppressDiagnostics` and consolidate it to `DiagnosticSuppression`.
Currently, we have only 2 usage of `TypeCheckExprFlags::SuppressDiagnostics`, and one of them is already under `DiagnosticSuppression`.

Reply:
@rintaro I think `DiagnosticSuppression::isEnabled` returning true should correspond to `ConstraintSystemFlags::SuppressDiagnostics` flag set in `ConstraintSystem` so we only have one source of input about this in the solver.

Reply:
I added another commit to consolidate `TypeCheckExprFlags::SuppressDiagnostics` with `DiagnosticSuppression`. Is this what you are suggesting?

Reply:
Yes, I think the way you changed it is okay, I just how that we wouldn't need to decouple it again later...

Reply:
I'm all in favor of consolidating them. I don't have an opinion on which one is 'better', I just think it's confusing when there are two ways of doing the same thing :)